src/container/include/container/container_rpc.hpp

```diff
@@ -0,0 +1,400 @@
+};
+
+template <>
+struct input_trait<ByteVector> {
+  static const bool is_supported = true;
+  static const InputType input_type = InputType::Bytes;
+};
+
+template <>
+struct input_trait<IntVector> {
+  static const bool is_supported = true;
+  static const InputType input_type = InputType::Ints;
+};
+
+template <>
+struct input_trait<FloatVector> {
+  static const bool is_supported = true;
+  static const InputType input_type = InputType::Floats;
+};
+
+template <>
+struct input_trait<DoubleVector> {
+  static const bool is_supported = true;
+  static const InputType input_type = InputType::Doubles;
+};
+
+template <>
+struct input_trait<SerializableString> {
+  static const bool is_supported = true;
+  static const InputType input_type = InputType::Strings;
+};
+
+template <class I>
+class Model {
+ public:
+  Model() : input_type_(input_trait<I>::input_type) {
+    static_assert(input_trait<I>::is_supported,
+                  "Model must be of a supported input type!");
+  }
+
+  virtual std::vector<std::string> predict(
+      const std::vector<I> inputs) const = 0;
+
+  InputType get_input_type() const { return input_type_; }
+
+ private:
+  const InputType input_type_;
+};
+
+// This is not thread safe
+class PerformanceTimer {
+ public:
+  static void start_timing() {
+    log_.str("");
+    last_log_ = Clock::now();
+  }
+
+  static void log_elapsed(const std::string tag) {
+    Clock::time_point curr_time = Clock::now();
+    long log_diff_micros =
+        std::chrono::duration_cast<std::chrono::microseconds>(curr_time -
+                                                              last_log_)
+            .count();
+    log_ << tag << ": " << std::to_string(log_diff_micros) << " us"
+         << ", ";
+    last_log_ = curr_time;
+  }
+
+  static std::string get_log() { return log_.str(); }
+
+ private:
+  static std::stringstream log_;
+  static Clock::time_point last_log_;
+};
+
+class RPC {
+ public:
+  RPC();
+  ~RPC();
+
+  // Disallow copy
+  RPC(const RPC&) = delete;
+  RPC& operator=(const RPC&) = delete;
+
+  // Default move constructor and assignment.
+  RPC(RPC&& other) = default;
+  RPC& operator=(RPC&& other) = default;
+
+  template <typename D>
+  void start(Model<Input<D>>& model, std::string model_name, int model_version,
+             std::string clipper_ip, int clipper_port) {
+    if (active_) {
+      throw std::runtime_error(
+          "Cannot start a container that is already started!");
+    }
+    active_ = true;
+    const std::string clipper_address =
+        "tcp://" + clipper_ip + ":" + std::to_string(clipper_port);
+    log_info_formatted(
+        LOGGING_TAG_CONTAINER,
+        "Starting container RPC with clipper ip: {} and port: {}", clipper_ip,
+        clipper_port);
+    serving_thread_ = std::thread(
+        [this, clipper_address, &model, model_name, model_version]() {
+          serve_model(model, model_name, model_version, clipper_address);
+        });
+    serving_thread_.detach();
```

Comment:
Don't detach the serving thread, join the thread in the RPC destructor. I know we do this in `RPCService`